### ISSUE TITLE
Signal 15 from nvidia-persistenced

### ISSUE THREAD

**Issue by paulshealy1** [NVIDIA/nvidia-docker#720]
I've installed the NVIDIA 384.125 driver on Ubuntu 16.04, plus nvidia-docker2, and I'm trying to enable the persistence daemon. It consistently stops with a signal 15. The complete syslogs are below. The kernel is 4.13.0-1014-azure. I don't see anything relevant in dmesg.

This also happens when I run it interactively. However, it does not happen if I run it interactively with _/usr/bin/nvidia-persistenced --verbose_.

The complete syslogs:
```
Apr 26 20:40:04 servername systemd[1]: Started ACPI event daemon.
Apr 26 20:40:08 servername systemd[1]: Configuration file /lib/systemd/system/nvidia-persistenced.service is marked executable. Please remove executable permission bits. Proceeding anyway.
Apr 26 20:40:08 servername systemd[1]: Starting NVIDIA Persistence Daemon...
Apr 26 20:40:08 servername nvidia-persistenced: Verbose syslog connection opened
Apr 26 20:40:08 servername nvidia-persistenced: Now running with user ID 116 and group ID 123
Apr 26 20:40:08 servername nvidia-persistenced: Started (3400)
Apr 26 20:40:08 servername nvidia-persistenced: device 44857:00:00.0 - registered
Apr 26 20:40:08 servername nvidia-persistenced: device 50479:00:00.0 - registered
Apr 26 20:40:08 servername nvidia-persistenced: device 56370:00:00.0 - registered
Apr 26 20:40:08 servername nvidia-persistenced: device 63091:00:00.0 - registered
Apr 26 20:40:08 servername nvidia-persistenced: Local RPC service initialized
Apr 26 20:40:08 servername systemd[1]: Started NVIDIA Persistence Daemon.
Apr 26 20:40:08 servername systemd[1]: Started NVIDIA Persistence Daemon.
Apr 26 20:40:08 servername nvidia-persistenced: Received signal 15
Apr 26 20:40:08 servername systemd[1]: Stopping NVIDIA Persistence Daemon...
Apr 26 20:40:08 servername nvidia-persistenced: Socket closed.
Apr 26 20:40:08 servername nvidia-persistenced: PID file unlocked.
Apr 26 20:40:08 servername nvidia-persistenced: PID file closed.
Apr 26 20:40:08 servername nvidia-persistenced: The daemon no longer has permission to remove its runtime data directory /var/run/nvidia-persistenced
Apr 26 20:40:08 servername nvidia-persistenced: Shutdown (3400)
Apr 26 20:40:08 servername systemd[1]: Stopped NVIDIA Persistence Daemon.```

**Comment by aaronp24** [comment]
Weird. Is systemd killing it for some reason?

What's in /lib/systemd/system/nvidia-persistenced.service on your system?

**Comment by paulshealy1** [comment]
How do I tell if systemd is killing it? I don't see anything related in the logs, but maybe I'm missing something.

Here's what I have in the system file:

[Unit]
Description=NVIDIA Persistence Daemon
Wants=syslog.target

[Service]
Type=forking
ExecStart=/usr/bin/nvidia-persistenced --user nvidia-persistenced --no-persistence-mode --verbose
ExecStopPost=/bin/rm -rf /var/run/nvidia-persistenced

**Comment by paulshealy1** [comment]
I did some testing, and this service file will work:

[Unit]
Description=NVIDIA Persistence Daemon
Wants=syslog.target

[Service]
Type=forking
ExecStart=/usr/bin/nvidia-persistenced --verbose
ExecStopPost=/bin/rm -rf /var/run/nvidia-persistenced

[Install]
WantedBy=multi-user.target

Though I would prefer to not run this as root.

**Comment by hanvinke** [comment]
I had a similar problem. Adding: _After=systemd-user-sessions.service_ in the [Unit] section solved it.

_[Unit]
Description=NVIDIA Persistence Daemon
Wants=syslog.target
After=systemd-user-sessions.service_

I also noted that the error message: ¨**The daemon no longer has permission to remove its runtime data directory /var/run/nvidia-persistenced**¨ has gone also.

**Comment by paulshealy1** [comment]
@hanvinke, that solved it for me. Thanks!

**Comment by evictor** [comment]
Stumbled upon this from search... I had same problem, commented out the `PIDFile` line in my config and now it works. Was using the config file here as basis: https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#power9-setup

My config now looks like:

```
[Unit]
Description=NVIDIA Persistence Daemon
Wants=syslog.target

[Service]
Type=forking
;PIDFile=/var/run/nvidia-persistenced/nvidia-persistenced.pid
Restart=always
ExecStart=/usr/bin/nvidia-persistenced --verbose
ExecStopPost=/bin/rm -rf /var/run/nvidia-persistenced

[Install]
WantedBy=multi-user.target
```

**Comment by mgsloan** [comment]
Awesome!

I've been spending hours figuring out why my Quadro M2200 was only listing eDP-1-1 and not the other outputs.  Curiously, it worked fine after I installed the `390.77` driver a couple weeks ago, I'm not sure what exactly broke it along the way.  After using `journalctl -b` to compare an old healthy boot with more recent boots, I noticed an inconsistency between the behavior of `nvidia-persistencd`.

Anyway, I copied [@evictor 's config, above](https://github.com/NVIDIA/nvidia-docker/issues/720#issuecomment-389721511) into  `/lib/systemd/system/nvidia-persistenced.service`, and the problem went away!  Woohoo.  Before this fix, my config had the same contents as @paulshealy1 's before the fix.
